activerecord/lib/active_record/migration.rb

```diff
@@ -553,21 +553,40 @@ def self.current_version
     # This class is used to verify that all migrations have been run before
     # loading a web page if <tt>config.active_record.migration_error</tt> is set to :page_load
     class CheckPending
-      def initialize(app)
+      def initialize(app, file_watcher: ActiveSupport::FileUpdateChecker)
         @app = app
-        @last_check = 0
+        @needs_check = true
+        @mutex = Mutex.new
+        @file_watcher = file_watcher
       end
 
       def call(env)
-        mtime = ActiveRecord::Base.connection.migration_context.last_migration.mtime.to_i
-        if @last_check < mtime
-          ActiveRecord::Migration.check_pending!(connection)
-          @last_check = mtime
+        @mutex.synchronize do
+          @watcher ||= @file_watcher.new(*watchable_args) do
+            @needs_check = true
+            ActiveRecord::Migration.check_pending!(connection)
+            @needs_check = false
+          end
+
+          if @needs_check
+            @watcher.execute
+          else
+            @watcher.execute_if_updated
+          end
         end
+
         @app.call(env)
       end
 
       private
+        def watchable_args
+          paths = Array(connection.migration_context.migrations_paths)
+          dirs_hash = paths.map do |dir|
+            [dir, ["rb"]]
+          end.to_h
+          [[], dirs_hash]
+        end
```

Comment:
Considering my `existing_migration_files` comment, this could be: `Array(connection.migration_context.migrations_paths).index_with([ "rb" ])`

(Though we'd need to require the core ext for `index_with` too)

Reply:
TIL `index_with`. Switched to that.